006-improving-configurability-of-kafka-listeners.md

```diff
@@ -0,0 +1,240 @@
+```
+
+The listeners will be configurable as array.
+Each listener will have its own unique name and will specify the port and type as required values.
+The port can be set to anything apart from `9091` (internal replication listener) and `9404` (Prometheus).
+A new type service will be introduced for the internal listeners designed for apps running inside the same Kubernetes cluster.
+Additionally, all the types from the existing external listeners will be supported as well.
+
+together with this change, I suggest to change the default value of the listener `tls` flag from `true` to `false`.
+The current situation, when TLS is enabled by default without using the `tls` field and needed to explicitly use the `tls: false` to disable TLS seems unintuitive and confusing.
+
+### Backwards compatibility
+
+The old format can be easily converted into the new format without any information loss.
+The example YAML below corresponds to the first example from the _Current situation_ section:
+
+```yaml
+    listeners:
+      - name: plain
+        port: 9092
+        type: service
+        tls: false
+        authentication:
+          type: scram-sha-512
+        networkPolicyPeers:
+          - podSelector:
+              matchLabels:
+                app: kafka-plaintext-consumer
+          - podSelector:
+              matchLabels:
+                app: kafka-plaintext-producer
+      - name: tls
+        port: 9093
+        type: service
+        tls: true
+        authentication:
+          type: tls
+        networkPolicyPeers:
+          - podSelector:
+              matchLabels:
+                app: kafka-consumer
+          - podSelector:
+              matchLabels:
+                app: kafka-producer
+      - name: external
+        port: 9094
+        type: route
+        tls: true
+        authentication:
+          type: tls
+```
+
+The second example could be converted like this:
+
+```yaml
+    listeners:
+      - name: plain
+        port: 9092
+        type: service
+      - name: tls
+        port: 9093
+        type: service
+```
+
+Being easily able to convert the old and new structures makes it easy to have single code in our operator while being able to easily work with the old format as well.
+
+### Code changes
+
+Since both the new and the old structure will use the same `listeners` property, the CRD generator will need to be enhanced to support the OpenAPI `oneOf` for different types.
+(The CRD generator already supports `oneOf` for having allowed only one of multiple fields defined, but not for having one field with multiple different types)
+
+```yaml
+listeners:
+  oneOf:
+    - type: object
+        properties:
+          # ...
+    - type: array
+        items:
+          # ...
+```
+
+In the `api` module, the setter and getter for the `listeners` field will use generic `JsonNode` type and inside the `api` module decide whether it should be decoded into the new or old format when the getter or setter is called.
```

Comment:
I think you could also use a `JsonSerializer` which would allow you to have a more strongly typed POJO with `listenerList` and `listenerObject` properties. You can see how that would work by looking at the fabric8 `IntOrString` class.

Reply:
I changed it to cover both options so that we can decide during implementation. I used the `JsonNode` in my prototype and know it can work. But maybe the custom (de)serializer is a bit cleaner and more elegant approach. So maybe we I try it later.